hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/JDBCSource.java

```diff
@@ -0,0 +1,239 @@
+        String value = properties.getString(prop);
+        if (!StringUtils.isNullOrEmpty(value)) {
+          LOG.info(String.format("Adding %s -> %s to jdbc options", key, value));
+          dataFrameReader.option(key, value);
+        } else {
+          LOG.warn(String.format("Skipping %s jdbc option as value is null or empty", key));
+        }
+      }
+    }
+  }
+
+  @Override
+  protected Pair<Option<Dataset<Row>>, String> fetchNextBatch(Option<String> lastCkptStr, long sourceLimit) {
+    try {
+      DataSourceUtils.checkRequiredProperties(props, Arrays.asList(Config.URL, Config.DRIVER_CLASS, Config.USER,
+          Config.RDBMS_TABLE_NAME, Config.IS_INCREMENTAL));
+      Option<String> lastCheckpoint =
+          lastCkptStr.isPresent() ? lastCkptStr.get().isEmpty() ? Option.empty() : lastCkptStr : Option.empty();
+      return fetch(lastCheckpoint);
+    } catch (Exception e) {
+      LOG.error("Exception while running JDBCSource ", e);
+      return Pair.of(Option.empty(), null);
+    }
+  }
+
+  @NotNull
+  private Pair<Option<Dataset<Row>>, String> fetch(Option<String> lastCheckpoint) {
+    Dataset<Row> dataset;
+    boolean isIncremental = props.getBoolean(Config.IS_INCREMENTAL);
+    if (lastCheckpoint.equals(Option.empty()) || StringUtils.isNullOrEmpty(lastCheckpoint.get())) {
+      LOG.info("No checkpoint references found... Doing a full rdbms table fetch");
+      dataset = fullFetch();
+    } else {
+      dataset = incrementalFetch(lastCheckpoint);
+    }
+    return Pair.of(Option.of(dataset), checkpoint(dataset, isIncremental));
+  }
+
+  @NotNull
+  private Dataset<Row> incrementalFetch(Option<String> lastCheckpoint) {
+    try {
+      final String ppdQuery = "(select * from %s where %s >= \"%s\") rdbms_table";
+      String query = String
+          .format(ppdQuery, props.getString(Config.RDBMS_TABLE_NAME), props.getString(Config.INCREMENTAL_COLUMN),
+              lastCheckpoint.get());
+      LOG.info(String
+          .format("Referenced last checkpoint and prepared new predicate pushdown query for jdbc pull %s", query));
+      return validatePropsAndGetDataFrameReader(sparkSession, props)
+          .option(Config.RDBMS_TABLE_PROP, query).load();
+    } catch (Exception e) {
+      LOG.error("Error while performing an incremental fetch... \n Note: Not all database support the PPD query we "
+          + "generate to do na incremental scan", e);
+      LOG.warn("Falling back to full scan.......");
+      return fullFetch();
+    }
+  }
+
+  private Dataset<Row> fullFetch() {
+    return validatePropsAndGetDataFrameReader(sparkSession, props).load();
+  }
+
+  private String checkpoint(Dataset<Row> rowDataset, boolean isIncremental) {
+    try {
+      if (isIncremental) {
+        Column incrementalColumn = rowDataset.col(props.getString(Config.INCREMENTAL_COLUMN));
+        final String max = rowDataset.agg(functions.max(incrementalColumn).cast(DataTypes.StringType)).first()
```

Comment:
this would make a second pass over `rowDataset` right.. Should we cache this so that we don't go and read from the database twice? Line 126 above, will do one fetch and compute the checkpoint and then the rowDataset is passed back to DeltaSync class back, and we start to fetch rows to write to Hudi, it will trigger another "recomputation" .. We should atleast support an option to cache this dataset IMO. 

Also is there a way to do this via Accumulators? (could be tricky) :D We can file a JIRA for later?

Reply:
Sure for now should we cache in the fetch() method, get the max of checkpointing, uncache and then return the df?

Reply:
yes. lets cache it in fetch() and add a parameter to control the persistence level, like we do in BloomIndex please

Reply:
> hoodie.datasource.jdbc.storage.level="MEMORY_ONLY_SER"
Has been added to the props file to get storage level. If no storage level is given by the user then we use the default as MEMORY_AND_DISK_SER